### PR TITLE
[framework upgrade] Refactor artifact generation and management for genesis

### PR DESCRIPTION
**Pull request by wrwg**
### Description

This is a larger refactoring regarding how we generate artifacts and releases, and how to embed them into the code. This refactoring is necessary to make upgradability for framework modules work.

> Background: until now our entire stack dealt with Move artifacts (.mv, .abi, etc.) in a way predating the introduction of packages (roughly 1 year ago). For example, to include code into genesis, we searched certain directories to collect all `.mv` files. Besides being outdated, this approach does not allow us to organize the genesis code so it is upgradable. Rather for this to work, we need to deploy code during genesis on package level instead of a bunch of `.mv` files.

Changes:
- A couple of new concepts have been introduced in `aptos-move/framework`:
  - a `ReleasePackage` for representing code and metadata of a package
  - a `ReleaseBundle` for representing a list of packages, in the order they need to be deployed
  - `ReleaseTarget` enum for distinguishing current, devnet, etc. release. The release target determines which framework packages to include (e.g. mainnet does not contain experimental packages)
- The creation of a framework release is now done over a small CLi (reachable as `cargo run -p framework -- release --target=<target>`). This command also generates/updates the SDK builders for framework and token.
- `cached-packages` is gone, all releated functionality has moved to the `framework` crate. There is no automatic on-the-fly compilation of the framework. Rather there are tests which ensure that the 'current' (head) framework artifacts are up-to-date.
- The dozen or so functions spread over framework/cached-packages to access module blobs, files, etc. are replaced by accessing a release bundle stored in a `static Lazy`.
- Genesis has been refactored to work with a `ReleaseBundle` instead of `Vec<Vec<u8>>`. Eventually, during genesis `code::initialize` is called to build the package metadata. This change cascades to all users of genesis, e.g. the CLI or aptos-node.
- Many places over the stack have been adapted to `ReleaseBundle` but not for all this is possible without a deeper refactor. For those legacy use cases, `ReleaseBundle` offers some helpers like `module_blobs`, `files`, etc. to simulate the old artifact APIs.

### Test Plan

There are two new tests in `aptos-move/e2e-move-tests/src/code_publishing.rs` which demonstrate that upgrading
of the Move stdlib (as one example) works.

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/2933)

### CODE REVIEW

File: crates/aptos/src/genesis/mod.rs [hunk 1]
```diff
@@ -131,13 +131,14 @@ pub fn fetch_genesis_info(git_options: GitOptions) -> CliTypedResult<GenesisInfo
         ));
     }
 
-    let modules = client.get_modules("framework")?;
+    // Should the target be configurable?
+    let framework = client.get_framework()?;
```
Comment (gregnazario) [inline crates/aptos/src/genesis/mod.rs:135]: We made it not configurable, considering it only happens once.

File: aptos-move/framework/src/generated/mod.rs [hunk 2]
```diff
@@ -0,0 +1,6 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+pub mod aptos_framework_sdk_builder;
+pub mod aptos_stdlib;
+pub mod aptos_token_sdk_builder;
```
Comment (gregnazario) [inline aptos-move/framework/src/generated/mod.rs:6]: Let's try to keep the refactors in separate commits next time.  Makes it easier to check the things that have proper new code changes.

File: aptos-move/framework/aptos-framework/sources/code.move [hunk 3]
```diff
@@ -147,8 +173,10 @@ module aptos_framework::code {
             let j = 0;
             while (j < vector::length(new_modules)) {
                 let name = vector::borrow(new_modules, j);
-                assert!(&old_mod.name != name, error::already_exists(EMODULE_NAME_CLASH))
-            }
+                assert!(&old_mod.name != name, error::already_exists(EMODULE_NAME_CLASH));
+                j = j + 1;
+            };
+            i = i + 1;
```
Comment (gregnazario) [inline aptos-move/framework/aptos-framework/sources/code.move:179]: I find it interesting this code had 2 for loops without adjusting the counters prior.

File: aptos-move/e2e-move-tests/src/harness.rs [hunk 4]
```diff
@@ -68,11 +66,16 @@ impl MoveHarness {
...
-        self.txn_seq_no.insert(addr, 10);
         self.executor.add_account_data(&data);
+        self.txn_seq_no.insert(addr, 10);
         data.account().clone()
     }
 
+    /// Gets the account for aptos root.
+    pub fn aptos_root_account(&mut self) -> Account {
+        self.new_account_at(AccountAddress::ONE)
+    }
```
Comment (gregnazario) [inline aptos-move/e2e-move-tests/src/harness.rs:77]: The Aptos root account used to be `0xA5501C`. has this changed?  There's already a function for that one

File: api/src/tests/test_context.rs [hunk 5]
```diff
@@ -229,6 +229,30 @@ impl TestContext {
...
     }
 
+    pub fn check_golden_output_pruned(&mut self, msg: Value) {
+        self.check_golden_output(Self::prune_golden(msg))
+    }
+
+    /// Prune well-known excessively large entries from a resource array response.
+    /// TODO: we can't dump all resources of an account as golden output. As functionality
+    /// grows this becomes too much. Need a way to filter only the resources which folks want.
+    fn prune_golden(val: Value) -> Value {
```
Comment (gregnazario) [inline api/src/tests/test_context.rs:239]: This kinda seems like the Golden files are being a pain if we're cutting these out.

Honestly it doesn't really matter all that much to me

File: aptos-move/framework/aptos-framework/sources/code.move [hunk 6]
```diff
@@ -84,6 +88,18 @@ module aptos_framework::code {
         from.policy <= to.policy
     }
 
+    /// Initialize package metadata for Genesis.
+    fun initialize(aptos_framework: &signer, package_owner: &signer, metadata: PackageMetadata)
+    acquires PackageRegistry {
```
Comment (gregnazario) [inline aptos-move/framework/aptos-framework/sources/code.move:93]: Feels like we'll need a formatter someday.  Not there's anything wrong with this format, but I've seen different elsewhere.

I would probably put these on the same line or indent the second line

File: aptos-move/vm-genesis/src/lib.rs [hunk 7]
```diff
@@ -211,7 +208,7 @@ fn exec_function(
                 "Error calling {}.{}: {}",
                 module_name,
                 function_name,
-                e.into_vm_status()
+                e // .into_vm_status()
```
Comment (gregnazario) [inline aptos-move/vm-genesis/src/lib.rs:211]: I'd delete this comment if it's no longer used.

File: aptos-move/framework/src/release_bundle.rs [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+        for path in &self.source_dirs {
+            let mut files = find_filenames(&[path], |p| extension_equals(p, MOVE_EXTENSION))?;
+            result.append(&mut files);
+        }
+        Ok(result)
+    }
+
+    pub fn named_addresses(&self) -> BTreeMap<String, NumericalAddress> {
+        BTreeMap::new()
+    }
```
Comment (gregnazario) [inline aptos-move/framework/src/release_bundle.rs:135]: Wondering, why is this a function, will it have addresses later?

File: aptos-move/framework/src/release_bundle.rs [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+}
+
+impl ReleasePackage {
+    /// Creates a new released package.
+    pub fn new(package: BuiltPackage) -> anyhow::Result<Self> {
+        Ok(ReleasePackage {
+            metadata: package.extract_metadata(/* TODO: remove */ UpgradePolicy::compat())?,
```
Comment (gregnazario) [inline aptos-move/framework/src/release_bundle.rs:142]: Should we remove?  What's the context?  It would be great to put in the comment.

**Review by gregnazario — COMMENTED** [review]
I mean it looks mostly fine.  Need to do a pass on all the package building.

Who's managing the releases?  Should there be a release version associated with it?

File: aptos-move/framework/src/aptos.rs [hunk 10]
```diff
@@ -3,89 +3,140 @@
...
+/// Returns the release bundle for the current code.
+pub fn current_release_bundle() -> &'static ReleaseBundle {
+    &CURRENT_RELEASE_BUNDLE
+}
 
-    dedup(framework, |f| f.self_id())
+/// Placeholder for returning the release bundle for the last devnet release(?).
+/// TODO: this is currently only used to differentiate between GenesisOptions::Fresh
+/// and GenesisOptions::Compiled. It is not clear what the difference should be.
+/// For now, we return the same as with current_release_bundel.
```
Comment (gregnazario) [inline aptos-move/framework/src/aptos.rs:139]: *bundle

File: aptos-move/framework/src/aptos.rs [hunk 11]
```diff
@@ -3,89 +3,140 @@
...
-    let mut files = super::move_files_in_path(MOVE_STDLIB_DIR);
-    files.extend(super::move_files_in_path(APTOS_STDLIB_DIR));
-    files.extend(super::move_files_in_path(APTOS_FRAMEWORK_DIR));
-    files.extend(super::move_files_in_path(TOKEN_MODULES_DIR));
-    files.extend(super::move_files_in_path(APTOS_STDLIB_DIR));
-    dedup(files, |f| f.to_string())
+impl Display for ReleaseTarget {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let str = match self {
+            ReleaseTarget::Current => "current",
```
Comment (gregnazario) [inline aptos-move/framework/src/aptos.rs:31]: Maybe we should rename `current` to `development` or `release-candidate` to show that it's not actually deployed anywhere.

File: aptos-move/framework/src/built_package.rs [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+    #[clap(long, default_value = "true")]
+    pub with_srcs: bool,
+    #[clap(long, default_value = "true")]
+    pub with_abis: bool,
+    #[clap(long, default_value = "true")]
+    pub with_source_maps: bool,
+    #[clap(long, default_value = "true")]
+    pub with_error_map: bool,
+    #[clap(skip)] // TODO: have a parser for this; there is one in the CLI buts its  downstream
+    pub named_addresses: BTreeMap<String, AccountAddress>,
```
Comment (gregnazario) [inline aptos-move/framework/src/built_package.rs:27]: Yeah, this is something that needs to be pulled out into a common CLI package

File: aptos-move/framework/src/built_package.rs [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+        }
+    }
+}
+
+/// Represents a built package.  It allows to extract `PackageMetadata`. Can also be used to
+/// just build Move code.
+pub struct BuiltPackage {
+    package_path: PathBuf,
+    #[allow(unused)]
+    options: BuildOptions,
```
Comment (gregnazario) [inline aptos-move/framework/src/built_package.rs:49]: Still unused?

File: aptos-move/framework/src/error_map.rs [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+    ) {
+        let mut gen = ErrmapGen::new(&model);
+        gen.gen();
+        Some(gen.finish())
+    } else {
+        None
+    }
+}
+
+struct ErrmapGen<'env> {
```
Comment (gregnazario) [inline aptos-move/framework/src/error_map.rs:54]: Nit, I'd probably name this `ErrMapGen`

File: aptos-move/framework/src/error_map.rs [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+    error_category_module: ModuleId,
+}
+
+impl<'env> ErrmapGen<'env> {
+    fn new(env: &'env GlobalEnv) -> Self {
+        Self {
+            env,
+            output: ErrorMapping::default(),
+            error_category_module: ModuleId::new(
+                AccountAddress::from_hex_literal("0x1").unwrap(),
```
Comment (gregnazario) [inline aptos-move/framework/src/error_map.rs:69]: AccountAddress::ONE

File: aptos-move/framework/src/generated/mod.rs (hunk 2 above)
Comment (wrwg) [reply aptos-move/framework/src/generated/mod.rs:6]: It is impossible for such a refactor, or it will take 10x more time. You can't design this top-down. There is so many legacy code which uses code and calls the Move compiler in some strange ways (mostly tests) that the design needed to be revised multiple times to satisfy all requirements.

File: aptos-move/e2e-move-tests/src/harness.rs (hunk 4 above)
Comment (wrwg) [reply aptos-move/e2e-move-tests/src/harness.rs:77]: But the framework is still deployed to 0x1. I need an account for 0x1 here because I upgrade framework code.

File: api/src/tests/test_context.rs (hunk 5 above)
Comment (wrwg) [reply api/src/tests/test_context.rs:239]: I think we should have a general concept in our REST APIs: *field masks*. These are quite common in REST Google APIs (disclaimer: I designed them personally 9 years ago so I might be biased, but it was a very common user request. [Here](https://developers.google.com/slides/api/guides/field-masks) is a gentle introduction.

Field masks is simply a set of field names which you use to prune a JSON resource object on server side (ideally they are passed down as deep as possible into to storage layer). Say you have resource with fields `x`, `y`, `z`. You can request this resource now and specify that you only want to see `x`. This applies to REST GET and REST LIST. It's easy to implement.

File: aptos-move/e2e-move-tests/src/harness.rs (hunk 4 above)
Comment (gregnazario) [reply aptos-move/e2e-move-tests/src/harness.rs:77]: Okay, let's call it `aptos_framework_account` to prevent confusion.

File: aptos-move/framework/src/built_package.rs (hunk 13 above)
Comment (wrwg) [reply aptos-move/framework/src/built_package.rs]: Still, but I now removed it.

File: aptos-move/framework/src/error_map.rs (hunk 14 above)
Comment (wrwg) [reply aptos-move/framework/src/error_map.rs]: Done

File: aptos-move/framework/src/error_map.rs (hunk 15 above)
Comment (wrwg) [reply aptos-move/framework/src/error_map.rs]: Done

File: aptos-move/framework/src/release_bundle.rs (hunk 8 above)
Comment (wrwg) [reply aptos-move/framework/src/release_bundle.rs:139]: I wasn't sure for what this is needed (there was formerly a `framework::named_addresses` and it is used in some tests. Some tests grab source files and compile them by bypassing the package systems. I'm surprised that it can be empty, but I guess lets keep it until we know. I added a TODO.

File: aptos-move/framework/src/release_bundle.rs (hunk 9 above)
Comment (wrwg) [reply aptos-move/framework/src/release_bundle.rs:146]: This need to be done in another PR. Basically this should become part of `Move.toml`.

File: aptos-move/vm-genesis/src/lib.rs (hunk 7 above)
Comment (wrwg) [reply aptos-move/vm-genesis/src/lib.rs]: Done

File: crates/aptos/src/genesis/mod.rs (hunk 1 above)
Comment (wrwg) [reply crates/aptos/src/genesis/mod.rs]: Done

**Review by wrwg — COMMENTED** [review]
Thank you for the review @gregnazario !

File: api/src/tests/test_context.rs (hunk 5 above)
Comment (davidiw) [reply api/src/tests/test_context.rs:239]: Could we unify this with https://github.com/aptos-labs/aptos-core/blob/bcc1bd18937b7c4c69a4b47c64413a92adea7ded/api/src/tests/test_context.rs#L226

File: aptos-move/framework/src/generated/mod.rs (hunk 2 above)
Comment (wrwg) [reply aptos-move/framework/src/generated/mod.rs:6]: Oh I misunderstood you. You meant multiple commits in one PR, and you are right, this can be done better. To be honest I haven't really figured out how to comfortably rebase with a series of commits, that is why I'm reluctant. At FB and at Google, you push a button.

**Review by davidiw — COMMENTED** [review]
Well I suspected one day the cached-framework would be deleted...

Considering the new devx, it would be great to add some sort of readme to the framework directory that indicates the new development flow.

Also would be great to have some sort of test that ensures that the committed `mrb` files are up to date.

Finally, I think you'll need to update the build flows for docker:  https://github.com/aptos-labs/aptos-core/blob/bcc1bd18937b7c4c69a4b47c64413a92adea7ded/docker/rust-all.Dockerfile#L149

No major concerns on the first reading. :ninja: :+1:

File: aptos-move/framework/src/built_package.rs [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+#[derive(Debug, Clone, Parser, Serialize, Deserialize)]
+pub struct BuildOptions {
+    #[clap(long, default_value = "true")]
+    pub with_srcs: bool,
+    #[clap(long, default_value = "true")]
+    pub with_abis: bool,
+    #[clap(long, default_value = "true")]
+    pub with_source_maps: bool,
+    #[clap(long, default_value = "true")]
+    pub with_error_map: bool,
```
Comment (davidiw) [inline aptos-move/framework/src/built_package.rs:25]: this won't work, bools cannot be passed in via cli... just trying now with the aptos command line and the publish flow.

Comment (wrwg) [reply aptos-move/framework/src/built_package.rs:25]: > Considering the new devx, it would be great to add some sort of readme to the framework directory that indicates the new development flow

We should really do this, but it need to postponed for some days. The docs folks need is on the CLU level. which I plan to write.

Comment (wrwg) [reply aptos-move/framework/src/built_package.rs:25]: > Also would be great to have some sort of test that ensures that the committed mrb files are up to date.

It's there!

Comment (wrwg) [reply aptos-move/framework/src/built_package.rs:25]: > Finally, I think you'll need to update the build flows for docker:

Oh boy. Is there someone who owns this docker code? What is the objective of this code? Is it in regular use so we need keep it working to land this PR? 

@geekflyer

https://github.com/aptos-labs/aptos-core/blob/bcc1bd18937b7c4c69a4b47c64413a92adea7ded/docker/rust-all.Dockerfile#L149

Comment (wrwg) [reply aptos-move/framework/src/built_package.rs:25]: [Oh Github UI ... I could *not* comment on the individual comments in the stream. So all replies landed on the last item in the stream.]

Comment (geekflyer) [reply aptos-move/framework/src/built_package.rs:25]: > > Finally, I think you'll need to update the build flows for docker:
> 
> Oh boy. Is there someone who owns this docker code? What is the objective of this code? Is it in regular use so we need keep it working to land this PR?
> 
> @geekflyer
> 
> https://github.com/aptos-labs/aptos-core/blob/bcc1bd18937b7c4c69a4b47c64413a92adea7ded/docker/rust-all.Dockerfile#L149

Those lines in the docker build files have been touched last ~5 months ago in:
https://github.com/aptos-labs/aptos-core/pull/72/files#diff-cde3368c11183b2f9fb91b7fa3063d604f3170c12afe1efb8b94dd607b979885R49-R55
and
https://github.com/aptos-labs/aptos-core/commit/aa4684d1f634d6b7698d508581068c4eb75ef0d6#diff-cde3368c11183b2f9fb91b7fa3063d604f3170c12afe1efb8b94dd607b979885R38

@davidiw  and @sherry-x  can probably best answer the objective of this code.
Personally I'd say say just remove them and apply the label `CICD:run-e2e-test` and hope things pass.

**Comment by wrwg** [comment]
> Oh one last thing, is there any discernible difference in performance of the tests

Nope not I can see. We can always optimize as we move forward by using the handy memoize crate.

File: aptos-move/framework/src/built_package.rs (hunk 16 above)
Comment (areshand) [reply aptos-move/framework/src/built_package.rs:25]: I think this code is mainly for running release to generate the mv files and put them into a target folder that is eventually used to generate the genesis.

File: aptos-move/framework/src/built_package.rs [hunk 17]
```diff
@@ -0,0 +1,137 @@
...
+use move_deps::move_core_types::errmap::ErrorMapping;
+use move_deps::move_package::compilation::compiled_package::CompiledPackage;
+use move_deps::move_package::BuildConfig;
+use serde::{Deserialize, Serialize};
+use std::collections::BTreeMap;
+use std::path::PathBuf;
+
+/// Represents a set of options for building artifacts from Move.
+#[derive(Debug, Clone, Parser, Serialize, Deserialize)]
+pub struct BuildOptions {
```
Comment (areshand) [inline aptos-move/framework/src/built_package.rs:17]: after this change, would the mv files auto-generated once the source move files change?

Comment (wrwg) [reply aptos-move/framework/src/built_package.rs:17]: The release bundle will *not* be autogenerated because this leads to a fragile build environment. Rather there is a test that everything is up-to-date. If that test fails, you have to regenerate manually.

**Review by davidiw — APPROVED** [review]
feel free to call out anything you want additional eyes on

**Review by geekflyer — APPROVED** [review]
